### PR TITLE
Llama 70B vLLM determinism fixes (bug fix)

### PR DESCRIPTION
**Pull request by ifabijanic-tt**
### PR Category
Bug fix

### Summary
This change fixes intermittent determinism failures in vLLM TT plugin tests for Llama 3.3 70B on Galaxy systems.

The failures were seen mostly in:

- `test_seeding_and_variety.py`
- `test_request_isolation.py`
- `test_tt_penalties.py`

Symptoms included seeded requests producing different outputs across runs, greedy requests occasionally diverging within the same batch, and penalty-enabled requests producing inconsistent results in mixed batches.

### Problem

vLLM treats batch slots as temporary execution rows, not stable request identities. Requests can be removed, re-added, compacted, or moved to different physical slots depending on scheduler decisions, prefill/decode grouping, CI timing, and batch composition.

The TT sampling path had several pieces of state that were effectively tied to physical slot position:

- request seed state
- seed counters
- RNG state
- sampling params
- penalty reset state
- decode trace/input state

That made deterministic behavior sensitive to vLLM scheduling layout. CI exposed this more frequently than local runs because request grouping and slot reuse patterns differ under load.

### Changes

- Slot-stable explicit seed handling

Explicit request seeds now derive per-token device seeds from a stable hash of:

```text
request_seed + token_counter
```

instead of depending only on mutable per-slot RNG progression.

This makes seeded sampling reproducible even if the request moves between physical slots.

- Seed counter and RNG remapping

When vLLM condenses the persistent batch, the TT seed manager now applies `slot_remap`.

```text
remap[new_slot] = old_slot
```

The seed manager moves seed values, seed counters, and RNG state from the old slot to the new slot.

This prevents a request from accidentally inheriting another slot’s sampling stream after compaction.

- Active-slot-only seed advancement

Decode now advances seed state only for active slots. Padded/inactive rows are ignored and receive sentinel values.

This prevents inactive rows from consuming seed steps or carrying stale state into later requests.

- Counter alignment to decode position

For explicit seeds, seed counters can be aligned to absolute decode positions.

This handles cases where vLLM temporarily removes/re-adds requests or resumes them in a different slot. The random stream is tied to the logical token position, not to how long the request happened to live in a physical slot.

- Greedy path normalization

Greedy requests are normalized consistently:

```text
temperature = 1
top_k = 1
top_p = 0
```

This keeps greedy rows on the force-argmax path and avoids accidental interaction with random/top-p sampling behavior.

- Safer prefill sampling

The Galaxy prefill path now avoids unsafe batched prefill behavior for explicit seeded cases.

Explicit seeded prefill requests are sampled through a canonical row-0 view while advancing the real request slot’s seed counter. This makes seeded prefill output independent of physical slot position and vLLM admission grouping.

- Decode input reset after prefill

On-device prefill sampling can switch the model into decode mode and leave cached decode inputs/traces stale.

A new `_decode_inputs_need_reset` flag forces the next decode after prefill to reload decode inputs once.

This addresses CI-only failures where the first decode after prefill could reuse stale inputs because the layout appeared unchanged.

- Sampling sub-core grid fix

`manual_seed` and `ttnn.sampling` now use the intended sampling sub-core grid.

This avoids sub-device/core mismatches and prevents sampling kernels from launching on an unexpected worker grid.

- Concat/sub-device ordering fix

The full-batch prefill logits concat is kept before switching to decode mode, avoiding the decode sub-device mismatch without using the sub-core concat path for the real 32-slot batch.

This addresses the fatal:

```text
Kernel group cores do not match sub device cores for programmable core type TENSIX
```

while avoiding slot/logit instability seen with the broader concat-grid workaround.

- Deterministic penalty reset

Penalty reset logic now builds prompt/output token count masks on host and copies deterministic tensors to device.

This removes device-side scatter-add during reset over duplicate-heavy prompts such as:

```text
a a a a a ...
a b c a b c ...
```

Those duplicate token updates were a likely source of nondeterministic penalty state in mixed penalty batches.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:llama_merged_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:llama_merged_fixes)

### CODE REVIEW

**Review by mtairum — APPROVED** [review]
Approving to unblock

**Comment by yieldthought** [comment]
Codex review:

[models/demos/llama3_70b_galaxy/tt/generator.py:797](https://github.com/tenstorrent/tt-metal/blob/2c817f336b489c1a14fb6f5a1e8210a3c07e105a/models/demos/llama3_70b_galaxy/tt/generator.py#L797): Mixed logprob seeded prefills return a shorter logprob tensor. In the new explicit_seeded_prefill path, log_prob_values only gets an entry when the current per-request sampling call returns tt_log_probs. If a seeded prefill batch mixes requests with logprobs disabled and enabled, prefill_log_probs becomes shorter than output_toks but prefill_forward_text still returns (output_toks, prefill_log_probs). The vLLM output path treats tuple logprobs as batch-aligned and indexes one logprob per sequence, so this can either attach logprobs to the wrong request or raise once it reaches a request past the shortened list. Suggestion: when any request in this explicit-seeded prefill needs logprobs, preserve request-order length and insert the same dummy 0.0 value used by the no-logprobs path for requests with logprobs disabled.

**Review by yieldthought — APPROVED** [review]
Approved subject to investigating and if valid fixing the codex issue

File: models/common/sampling/generator.py [hunk 1]
```diff
@@ -18,6 +18,27 @@
...
+
+    The device sampling op accepts bounded positive seeds, while vLLM
+    request seeds can be any integer and must be reproducible regardless
+    of batch slot. Hashing (request seed, token counter) gives each token
+    a deterministic but well-mixed device seed without relying on mutable
+    per-slot RNG state. The constants below are the SplitMix64 finalizer.
+    """
+    value = (int(seed) & _UINT64_MASK) ^ (
+        (int(counter) + 0x9E3779B97F4A7C15) & _UINT64_MASK
+    )
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/common/sampling/generator.py]: What is the purpose of this? Only get hash?

**Comment by djordje-tt** [comment]
*Claude-assisted review under Djordje Ivanovic's supervision.*

Good fix overall — the root cause analysis is solid and the slot-stable seed hashing approach is the right architectural direction. A few items worth addressing:

---

**1. Leftover commented-out print — `models/demos/llama3_70b_galaxy/tt/generator.py`**

The non-seeded prefill branch retains a commented-out debug print that was already present before this PR:
```python
# print("sampling_params_scattered", sampling_params, "empty_slots", empty_slots)
```
It moved from the pre-existing path into the new `else` branch but was not removed. Should be cleaned up.

---

**2. `_scatter_params_to_slots` filler index fix is correct but the old bug deserves a note**

Old code: `filler = values[len(slots)] if len(values) > len(slots) else values[-1]`
New code: `filler = user_vals[-1] if user_vals else values[-1]`

The old code silently fell back to `values[-1]` when `len(values) == len(slots)` (common case), which is the same as `user_vals[-1]`. But when `len(values) > len(slots)` the old code used `values[len(slots)]` — a padding element, not the last active slot's value. The new code is more predictable. Consider a short comment explaining the invariant ("filler is the last active request's value, not a padding element").

---

**3. `_seed_from_slot_params` out-of-bounds fallback — `models/common/sampling/generator.py`**

When `slot >= flat.numel()` (tensor) or `slot >= len(seeds)` (list), the method returns `seeds[0]` / `flat[0]`. This silently assigns slot 0's seed to an out-of-bounds slot instead of returning `None`. In practice this shouldn't be reached during normal vLLM operation, but if it is (e.g. during a batch condense edge case), the incorrect seed could cause a seeded request to appear non-seeded or to share state with another request. Returning `None` for out-of-bounds slots would be safer.

Same pattern exists in `align_seed_counters_to_positions` for the list-positions case: `positions[slot] if len(positions) > slot else positions[0]` — out-of-bounds slots fall back to `positions[0]` instead of being skipped.

---

**4. ⚠️ `ttnn.synchronize_device` on every decode step for seeded requests — `models/common/sampling/generator.py` — please clarify**

```python
if self._active_request_seed:
    ttnn.synchronize_device(self.tt_sampling.mesh_device)
```

This is a hard blocking stall injected into every single decode token step for any batch that contains at least one explicitly seeded request. On Galaxy this means every token generation for a seeded request pays a full host-device sync round-trip.

A few questions:

- **What specific bug does this synchronize call fix?** Is there a documented race condition between the seed tensor write and the device reading it in the sampling op? Or was it added defensively/empirically to make a flaky test pass?
- **Is it actually necessary?** `ttnn.copy_host_to_device_tensor` already issues the copy command. If the subsequent sampling kernel is enqueued into the same command queue after the copy, the device should serialize them automatically without an explicit host-side sync. A host-side `synchronize_device` only makes sense here if either (a) the seed tensor write and the sampling op are on different queues/devices, or (b) there is a real observed ordering problem that the queue alone doesn't handle.
- **What is the measured throughput impact?** Even a 1–2 ms sync stall per token on Galaxy at short decode lengths is significant relative to token generation time.
- **Is there a better fix?** If the issue is that the seed tensor needs to be visible before the next op, ensuring both are enqueued into the same TTNN command queue (without a host sync) should be sufficient and have zero throughput penalty. If the issue is something else entirely, please document it so this isn't accidentally removed later — or kept longer than necessary.

---

**5. Seeded prefill peak device memory — `models/demos/llama3_70b_galaxy/tt/generator.py`**

The seeded prefill loop does:
```python
single_logits_batch = ttnn.concat([logits_source[slot]] * max_batch, dim=2, ...)
```
for each seeded request, allocating `max_batch` copies of a logit shard. With `max_batch=32` and large vocab shards on Galaxy, this could be non-trivial in peak device memory, particularly if multiple seeded requests are admitted in the same prefill batch. The `ttnn.deallocate(single_logits_batch)` helps, but peak is still one allocation per iteration. Worth verifying against worst-case memory budget.

---

**6. `broadcast_sampling_params` fix — correct, worth a comment**

The old logic had a subtle bug: a list element that happened to be `None` would fall through to the `if chosen is None` branch and set `kwargs[f.name] = None` instead of `[None] * slot_len`. The fix (tracking `value_is_list` separately) is correct. A one-line comment here would make the invariant explicit for future readers.

---

**7. `top_p[i] = 0.0` for greedy slots — verify device kernel boundary**

`_normalise_and_pad` now sets `top_p[i] = 0.0` when `temperature[i] == 0`. `_is_force_argmax_sampling` in `tt_sampling.py` was updated to treat `p=0.0` as equivalent to `p=1.0` for the force-argmax fast path. Before this lands, confirm the device-side top_p kernel also treats `top_p=0` as "pass all tokens" (not "pass no tokens"), since `0.0` could be interpreted either way depending on the kernel's boundary condition.

File: models/common/sampling/generator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: Yes, this is a deterministic hash/mixing step. We combine the request seed with the per-token counter, then run it through the SplitMix64 finalizer below. The goal is to derive a stable, well-mixed device seed for each token from (request_seed, token_counter) and to avoid reusing the same device seed on every decode step. That should (in theory and actually in practice for a lot of cases) avoid correlated sequential seeds, and keeps seeded sampling reproducible even if the request moves between batch slots. The masks keep the math in uint64 wraparound semantics. Finally we map the result into the bounded device seed range with return (value % DEVICE_SEED_MAX) + 1.

**Comment by ifabijanic-tt** [comment]
Added changes from code reviews

**Comment by mtairum** [comment]
@ifabijanic-tt there's still a conflict with generator.py. Needs rebasing.

**Review by djordje-tt — APPROVED** [review]

**Comment by djordje-tt** [comment]
Ublocked merge, lgtm just please run pipelines (vllm nightly, tier1 models at least, perf pipelines). To cover that there are no regressions, nor failures.
